File: tasks/translation/output/Qwen/Qwen2.5-32B-Instruct-AWQ/codenet/cot/Java/Go/s584572995_commented.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

const mod = 1e9 + 7
var to [][]int
var dp []int

func main() {
	fs := NewFastScanner(os.Stdin)
	N := fs.NextInt()
	to = make([][]int, N)
	for i := 0; i < N; i++ {
		to[i] = make([]int, 0)
	}
	edges := make([][2]int, N-1)
	for i := 0; i < N-1; i++ {
		A := fs.NextInt() - 1
		B := fs.NextInt() - 1
		to[A] = append(to[A], B)
		to[B] = append(to[B], A)
		edges[i][0] = B
		edges[i][1] = A
	}
	dp = make([]int, N)
	rec(0, -1)
	mulsum := 0
	for _, e := range edges {
		candidate1 := e[0]
		candidate2 := e[1]
		child := candidate1
		if dp[candidate1] > dp[candidate2] {
			child = candidate2
		}
		a := dp[child]
		b := N - a
		mul := (modpow(2, a) - 1) * (modpow(2, b) - 1) % mod
		mulsum += mul
		mulsum %= mod
	}
	cases := modpow(2, N)
	PN := (mulsum + (cases - 1 + mod) % mod) % mod
	BN := N * modpow(2, N-1) % mod
	WN := (PN - BN + mod) % mod
	ans := WN * modpow(cases, mod-2)
	ans %= mod
	fmt.Println(ans)
}

func rec(v int, parent int) int {
	res := 1
	for _, next := range to[v] {
		if next == parent {
			continue
		}
		res += rec(next, v)
	}
	dp[v] = res
	return res
}

func modpow(x int, y int) int {
	if y == 0 {
		return 1
	}
	if y%2 != 0 {
		return x * modpow(x, y-1) % mod
	}
	tmp := modpow(x, y/2)
	return tmp * tmp % mod
}

type FastScanner struct {
	reader *bufio.Reader
}

func NewFastScanner(r io.Reader) *FastScanner {
	return &FastScanner{reader: bufio.NewReader(r)}
}

func (fs *FastScanner) Next() string {
	s, _ := fs.reader.ReadString(' ')
	return strings.TrimSpace(s)
}

func (fs *FastScanner) NextInt() int {
	s := fs.Next()
	n, _ := strconv.Atoi(s)
	return n
}

// End of Code

File: tasks/translation/output/Qwen/Qwen2.5-32B-Instruct-AWQ/codenet/cot/Java/C/s584572995_commented.c
#include <stdio.h>
#include <stdlib.h>

#define mod 1000000007
int N;
int to[100000];
int dp[100000];
int edges[100000][2];

int rec(int v, int parent) {
    int res = 1;
    for (int i = 0; i < to[v]; i++) {
        int next = to[v * 100000 + i];
        if (next == parent) continue;
        res += rec(next, v);
    }
    return dp[v] = res;
}

long modpow(long x, long y) {
    if (y == 0) return 1;
    if (y % 2 != 0) return x * modpow(x, y - 1) % mod;
    long tmp = modpow(x, y / 2);
    return tmp * tmp % mod;
}

int main() {
    int A, B;
    scanf("%d", &N);
    for (int i = 0; i < N; i++) to[i] = 0;
    for (int i = 0; i < N - 1; i++) {
        scanf("%d %d", &A, &B);
        A--; B--;
        to[A * 100000 + to[A]] = B;
        to[B * 100000 + to[B]] = A;
        edges[i][0] = B;
        edges[i][1] = A;
        to[A]++; to[B]++;
    }
    for (int i = 0; i < N; i++) dp[i] = 0;
    rec(0, -1);
    long mulsum = 0;
    for (int i = 0; i < N - 1; i++) {
        int candidate1 = edges[i][0];
        int candidate2 = edges[i][1];
        int child = candidate1;
        if (dp[candidate1] > dp[candidate2]) child = candidate2;
        long a = dp[child];
        long b = N - a;
        long mul = (modpow(2, a) - 1) * (modpow(2, b) - 1) % mod;
        mulsum += mul;
        mulsum %= mod;
    }
    long cases = modpow(2, N);
    long PN = (mulsum + (cases - 1 + mod) % mod) % mod;
    long BN = N * modpow(2, N - 1) % mod;
    long WN = (PN - BN + mod) % mod;
    long ans = WN * modpow(cases, mod - 2) % mod;
    printf("%ld\n", ans);
    return 0;
}
// End of Code
